Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -1178,13 +1180,32 @@ void RenderBlockFlow::determineLogicalLeftPositionForChild(RenderBox& child, App
         startPosition += (style().isLeftToRightDirection() ? 1 : -1) * verticalScrollbarWidth();
     if (style().scrollbarGutter().bothEdges && !isHorizontalWritingMode())
         startPosition += (style().isLeftToRightDirection() ? 1 : -1) * horizontalScrollbarHeight();
-    LayoutUnit totalAvailableLogicalWidth = borderAndPaddingLogicalWidth() + availableLogicalWidth();
+    const LayoutUnit totalAvailableLogicalWidth = borderAndPaddingLogicalWidth() + availableLogicalWidth();
+
+    auto computeSelfAlignedLogicalLeftPositionForChild = [&](LayoutUnit currentPosition) -> LayoutUnit {
+        auto startAlignedPosition = currentPosition;
+        auto endAlignedPosition = totalAvailableLogicalWidth - startAlignedPosition - logicalWidthForChild(child);
+        auto selfAlignmentStyle = child.style().resolvedJustifySelf(&style(), ItemPosition::Normal);
+        auto resolvedDirection = selfAlignmentStyle.isSelfDirectioned() ? child.style().direction() : style().direction();
+        if (!isLeftToRightDirection(resolvedDirection))
+            std::swap(startAlignedPosition, endAlignedPosition);
+        if (child.style().justifySelf().overflow() == OverflowAlignment::Safe && logicalWidthForChild(child) > totalAvailableLogicalWidth)
+            return startAlignedPosition;
+        if (selfAlignmentStyle.isStartward(resolvedDirection) || selfAlignmentStyle.position() == ItemPosition::Stretch)
```

Comment:
`isStartward()` is already checking for stretch, right? So we don't need to do it here.